### PR TITLE
Add cpu.core_count and cpu.hyperthreading facts. Closes #355.

### LINKED ISSUE #355: cpu.cores Does not Always properly Account for Hyperthreading
# Openning a new issue
___

## Specify type:
- Bug

### Bug severity (if applicable):
- High
___

## Description:
For some earlier version of RHEL the number of CPU cores determined by the cpu.cores fact is 2x the number of physical cores due to hyperthreading. 

The bolded part of the command below shows when cpu.cores is not found in /proc/cpuinfo we just count the number of time "processor" appears in /proc/cpuinfo which is equivalent to logical CPUs which include hyperthreading. 

CPU_CORES=$(grep cpu.cores /proc/cpuinfo | sed 's/cpu.*:.\?//g' | sort -u); **if [ -z "$CPU_CORES" ]; then CPU_CORES=$(grep processor /proc/cpuinfo | wc -l);** else CPU_SOCKETS=$(grep "physical id" /proc/cpuinfo | sort -u| wc -l); let CPU_CORES=CPU_CORES\*CPU_SOCKETS; fi; echo "$CPU_CORES" | sed 's/\s*//g' 

___

### Version of rho:
0.28

### Expected behavior:
Should check if hyperthreading is enabled and if the cpu.cores is not found in /proc/cpuinfo then divide CPU cores by 2. 

Use the following to check if hyperthreading is enabled. 
number of siblings = 2 * number of cores per socket
cat /proc/cpuinfo |egrep 'sibling|cores' 

### Steps to reproduce:
Run against system with kernel 2.6.18-419.el5. 

### Environment information:
Provide detailed information on the *scanning* and *target* systems (if applicable); see example below.

Type            | Operating System | OS Version | Python Version | Virtualization
--------------- | ---------------- | ---------- | -------------- | --------------
Target System   | RHEL             | 2.6.18-419.el5

### PR DESCRIPTION
**Pull request by chambridge**

### CODE REVIEW

**Review by elyezer — APPROVED** [review]
Looks good.

I will make sure to include these new facts in the change I am about to do on the Camayoc's fact list

File: roles/cpu/tasks/main.yml [hunk 1]
```diff
@@ -71,37 +71,118 @@
...
+  when: '"cpu.hyperthreading" in facts_to_collect or "cpu.core_count" in facts_to_collect'
+
+- name: extract or default cpu.siblings fact
+  set_fact:
+    cpu_siblings: "{{ cpu_siblings_cmd['stdout_lines'][0] | default('Not Found')}}"
+  when: '"cpu.hyperthreading" in facts_to_collect or "cpu.core_count" in facts_to_collect'
+
+- name: determine cpu.hyperthreading fact
+  set_fact:
+    cpu_hyperthreading: "{{ cpu_siblings != 'Not Found' and cpu_siblings == cpu_core_per_socket }}"
```
Comment (mdvickst) [inline roles/cpu/tasks/main.yml:114]: I think the logic on the second condition is backwards maybe? 

If hyperthreading is enabled (only for RHEL on physical hosts) the number of siblings will be 2x the number of cores. So if cores is 4 and siblings is 8 then that would mean hyperthreading is enabled but I believe this fact would be set to False.

File: roles/cpu/tasks/main.yml [hunk 2]
```diff
@@ -71,37 +71,118 @@
...
   when: '"cpu.socket_count" in facts_to_collect'
+
+- name: calculate cpu.core_count
+  set_fact:
+    cpu_core_count: "{{ (cpu_core_per_socket | int) *  (cpu_socket_count | int)}}"
+  when: 'cpu_core_per_socket is defined and cpu_core_per_socket != "Not Found"'
+
+- name: calculate cpu.core_count fallback no hyperthreading
+  set_fact:
+    cpu_core_count: "{{ (cpu_count | int) *  (cpu_socket_count | int)}}"
```
Comment (mdvickst) [inline roles/cpu/tasks/main.yml:170]: I don't think this logic is correct. cpu_count will already be the total number of cores if hyperthreading is disabled so multiplying the cpu_count by the cpu_scoket_count will double or quadruple the cpu.core_count depending on how many sockets there are. 

I think this should just be equal to cpu.count if hyperthreading is disabled.

**Review by mdvickst — COMMENTED** [review]
Looks great except I think the logic might be off in two places.

**Comment by chambridge** [comment]
will reopen this issue to make updates.

**Pull request merged**